File: leetcode/897.cpp
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};

class Solution {
private:
  TreeNode *bst(TreeNode *node, TreeNode *head) {
    if (node == nullptr) {
      return head;
    }
    head = bst(node->left, head);
    node->left = nullptr;
    head->right = node;
    head = head->right;
    head = bst(node->right, head);
    return head;
  }

public:
  TreeNode *increasingBST(TreeNode *root) {
    TreeNode head;
    bst(root, &head);
    return head.right;
  }
};

File: leetcode/897.py
# Definition for a binary tree node.
class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:

    def increasingBST(self, root: TreeNode) -> TreeNode:

        def dfs(root: TreeNode):
            nonlocal head
            nonlocal cur
            if not root:
                return None
            dfs(root.left)
            if not head:
                head = root
            if not cur:
                cur = root
            else:
                root.left = None
                cur.right = root
                cur = cur.right
            dfs(root.right)

        head, cur = None, None
        dfs(root)
        return head